Make queue length configurable through the dataflow file

phil-opp opened the pull request:
Open questions:

- How do we want to name the config option? Is `max_queue_len` okay for now?
- Do we want to configure the queue length per node or per input?
- What should be the default queue length?

Based on #230

haixuanTao commented:
> - How do we want to name the config option? Is max_queue_len okay for now?
I feel: `queue_size` would be more appropriate. 

> Do we want to configure the queue length per node or per input?
Although, having a queue per input sound very interesting. It might be really useful for things that are really spammy such as `position`, `tick`, where most of the time the last one might be the only really useful one. 

Although the implementation might not be easy, knowing that you might have to recompute the order of the receiving messages depending on which one has been dropped.

> What should be the default queue length?

So, I feel like in general we should try to dimension to something like dora-drives. Characteristics are the following:
-> As input we would probably have ~ 10 sensors
-> sending at a ~ 10 Hz. 
-> We might have about ~10 operators.

In light of this, a default queue of length 10 would mean that at max we would have a 1s old message ( 10 * freq ). So I think that 10 was a fair default queue length.

phil-opp commented:
> I feel: `queue_size` would be more appropriate.

Agreed!



> Although, having a queue per input sound very interesting. It might be really useful for things that are really spammy such as `position`, `tick`, where most of the time the last one might be the only really useful one.
> 
> Although the implementation might not be easy, knowing that you might have to recompute the order of the receiving messages depending on which one has been dropped.

I changed the implementation to allow configuring the queue size per input. The syntax for this is:

```yaml
nodes:
  - id: webcam
    custom:
      source: ./webcam.py
      inputs:
        tick:
          source: dora/timer/millis/2
          queue_size: 1000
      outputs:
        - image
```

The old `tick: dora/timer/millis/2` format is still supported for convenience and backwards compatibility. The addional fields are filled with default values in this case.

I also updated the event drop code to be more efficient even for large queue sizes.

phil-opp commented on binaries/daemon/src/listener/mod.rs:
```diff
@@ -26,6 +30,7 @@ pub async fn spawn_listener_loop(
     node_id: &NodeId,
     daemon_tx: &mpsc::Sender<Event>,
     config: DaemonCommunicationConfig,
+    max_queue_len: BTreeMap<DataId, usize>,
```
Thanks, fixed in https://github.com/dora-rs/dora/pull/231/commits/ffb62b591b23de3ba437934ed1505547553cd5de.

phil-opp commented:
I pushed a new commit to consider the queue sizes for operator inputs too.
